### PR TITLE
scripts: release: list_backports.py

### PR DESCRIPTION
**Pull request by cfriedt**
* create `list_backports.py` to examine prs applied to a backport branch and extract associated issues
  * helpful for adding to release notes
* add github action to verify backported changes have an associated issue (CI workflow added)
* use `GITHUB_TOKEN` in `bug_bash.py` and `list_issues.py`
* change `bug_bash.py` to use more consistent variable names (e.g. `--end END_DATE`)

```
Usage:
    # generate release notes for v2.7, with a date range, excluding specific PRs, output ReST
    ./scripts/release/list_backports.py \
        -t ~/.ghtoken \
        -b v2.7-branch \
        -s 2021-12-15 -e 2022-04-22 \
        -P 45074 -P 45868 -P 44918 -P 41234 -P 41174

    # verify a specific PR to v3.0 has an associated issue, output JSON
    GITHUB_TOKEN="<secret>" \
    ./scripts/release/list_backports.py \
        -b v3.0-branch \
        -p 43381 \
        -j | jq . | tee /tmp/backports.json
```

### CODE REVIEW

File: scripts/release/bug_bash.py [hunk 1]
```diff
@@ -26,39 +26,39 @@ def parse_args():
     parser.add_argument('-a', '--all', dest='all',
                         help='Show all bugs squashed', action='store_true')
     parser.add_argument('-t', '--token', dest='tokenfile',
-                        help='File containing GitHub token', metavar='FILE')
-    parser.add_argument('-b', '--begin', dest='begin', help='begin date (YYYY-mm-dd)',
+                        help='File containing GitHub token (altnernatively, use GH_TOKEN env variable)', metavar='FILE')
```
Comment (stephanosio) [inline scripts/release/bug_bash.py:29]: typo

```suggestion
                        help='File containing GitHub token (alternatively, use GH_TOKEN env variable)', metavar='FILE')
```

File: scripts/release/bug_bash.py [hunk 2]
```diff
@@ -26,39 +26,39 @@ def parse_args():
...
     if args.tokenfile:
         with open(args.tokenfile, 'r') as file:
             token = file.read()
             token = token.strip()
     else:
-        if 'GITHUB_TOKEN' not in os.environ:
+        if 'GH_TOKEN' not in os.environ:
             raise ValueError('No credentials specified')
-        token = os.environ['GITHUB_TOKEN']
+        token = os.environ['GH_TOKEN']
```
Comment (stephanosio) [inline scripts/release/bug_bash.py:48]: Any reason behind this change? FWIW, GitHub calls it `GITHUB_TOKEN` (`${{ secrets.GITHUB_TOKEN }}`) and there are other scripts that use this name as well (see https://github.com/zephyrproject-rtos/zephyr/pull/45540).

Comment (cfriedt) [reply scripts/release/bug_bash.py:48]: Ah weird - I thought I saw another script that used this convention. Have not been doing ci much lately

Comment (stephanosio) [reply scripts/release/bug_bash.py:48]: https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/scripts/release/list_issues.py#L150

right, there is `list_issues.py` that uses `GH_TOKEN`. Could you instead update this file to use `GITHUB_TOKEN` here?

**Comment by cfriedt** [comment]
* accumulate PRs to ignore with `-x / --exclude-pull` option
* order pulls numerically in report

**Comment by cfriedt** [comment]
* fix typo
* correct `GH_TOKEN` to `GITHUB_TOKEN` in `list_issues.py`

File: scripts/release/bug_bash.py [hunk 3]
```diff
@@ -26,17 +26,17 @@ def parse_args():
     parser.add_argument('-a', '--all', dest='all',
                         help='Show all bugs squashed', action='store_true')
     parser.add_argument('-t', '--token', dest='tokenfile',
-                        help='File containing GitHub token', metavar='FILE')
-    parser.add_argument('-b', '--begin', dest='begin', help='begin date (YYYY-mm-dd)',
-                        metavar='date', type=valid_date_type, required=True)
+                        help='File containing GitHub token (altnernatively, use GITHUB_TOKEN env variable)', metavar='FILE')
```
Comment (stephanosio) [inline scripts/release/bug_bash.py:29]:
```suggestion
                        help='File containing GitHub token (alternatively, use GITHUB_TOKEN env variable)', metavar='FILE')
```

**Review by stephanosio — DISMISSED** [review]

File: scripts/release/bug_bash.py [hunk 4]
```diff
@@ -26,17 +26,17 @@ def parse_args():
     parser.add_argument('-a', '--all', dest='all',
                         help='Show all bugs squashed', action='store_true')
     parser.add_argument('-t', '--token', dest='tokenfile',
-                        help='File containing GitHub token', metavar='FILE')
-    parser.add_argument('-b', '--begin', dest='begin', help='begin date (YYYY-mm-dd)',
-                        metavar='date', type=valid_date_type, required=True)
+                        help='File containing GitHub token (alternatively, use GITHUB_TOKEN env variable)', metavar='FILE')
```
Comment (mbolivar-nordic) [inline scripts/release/bug_bash.py:29]: These changes really feel like they should be part of a separate commit.

File: scripts/release/list_issues.py [hunk 5]
```diff
@@ -147,10 +147,10 @@ def parse_args():
 def main():
     parse_args()
 
-    token = os.environ.get('GH_TOKEN', None)
+    token = os.environ.get('GITHUB_TOKEN', None)
```
Comment (mbolivar-nordic) [inline scripts/release/list_issues.py:150]: These changes are not mentioned in the commit log and also seem like they should be in their own commit.

Comment (cfriedt) [reply scripts/release/list_issues.py:150]: I was asked to include them here by Stephanos. Would be fine to mention it in the commit msg.

Comment (mbolivar-nordic) [reply scripts/release/list_issues.py:150]: I see that the change was requested, but I see no request (or reason for one) that it has to be in this same commit; is there one?

**Review by stephanosio — DISMISSED** [review]

File: scripts/release/list_backports.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    bp = Backport(Github(args.token), args.base,
+                  args.start, args.end, set(args.excludes))
+
+    pulls_without_issues = bp.get_pulls_without_issues()
+    if pulls_without_issues:
+        s = 'Please ensure the body of each PR to a release branch contains "Fixes #1234"\n'
+        s += 'The following PRs are lacking associated issues:'
+        for p in pulls_without_issues:
+            s += f'\nhttps://github.com/zephyrproject-rtos/zephyr/pull/{p.number}'
+        raise ValueError(s)
```
Comment (fabiobaltieri) [inline scripts/release/list_backports.py:209]: Would it make sense to use a print and explicit exit? It's a bit odd to see a backtrace when this kicks in but it may be just me... up to you really.

File: scripts/release/list_backports.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+        self.get_pulls()
+
+        return self._pulls_without_an_issue
+
+
+def main():
+    args = parse_args()
+    bp = Backport(Github(args.token), args.base,
+                  args.start, args.end, set(args.excludes))
```
[the remark this answers is not in the document]
Comment (cfriedt) [reply scripts/release/list_backports.py]: Fixed!

**Comment by cfriedt** [comment]
@stephanosio - PTAL at `.github/workflows/backport-has-issue.yml`

File: .github/workflows/backport-has-issue.yml [hunk 8]
```diff
@@ -0,0 +1,21 @@
+name: Backport has Issue
+on:
+  pull_request_target:
+    on:
+      push:
+        branches:
```
Comment (stephanosio) [inline .github/workflows/backport-has-issue.yml:6]: This does not seem right. Have you tested this?

AFAIK, `branches` should be directly under `pull_request_target`.

File: scripts/release/list_backports.py [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+    setattr(args, 'token', token)
+
+    return args
+
+
+class Backport(object):
+    def __init__(self, gh, base, pulls):
+        self._gh = gh
+        self._base = base
+        self._repo = gh.get_repo('zephyrproject-rtos/zephyr')
```
Comment (stephanosio) [inline scripts/release/list_backports.py:120]: the org and repo names need to be configurable via the script parameters so that this script and workflow can be reused in the forks.

also they should be explicitly specified in the workflow based on the execution context:
https://github.com/zephyrproject-rtos/zephyr/blob/51c7777529a1a28b1e5ab5569cf6bf0cea7bd2df/.github/workflows/assigner.yml#L35-L36

File: .github/workflows/backport-has-issue.yml [hunk 10]
```diff
@@ -0,0 +1,21 @@
+name: Backport has Issue
```
Comment (stephanosio) [inline .github/workflows/backport-has-issue.yml:1]: Please name this "Backport Issue Check" or something along the line.

**Review by stephanosio — DISMISSED** [review]

File: scripts/release/list_backports.py (hunk 9 above)
Comment (cfriedt) [reply scripts/release/list_backports.py]: Fixed!

File: .github/workflows/backport-has-issue.yml (hunk 10 above)
Comment (cfriedt) [reply .github/workflows/backport-has-issue.yml]: Fixed!

File: .github/workflows/backport-has-issue.yml (hunk 8 above)
Comment (cfriedt) [reply .github/workflows/backport-has-issue.yml]: Fixed!

Comment (cfriedt) [reply .github/workflows/backport-has-issue.yml]: Fixed!

**Comment by stephanosio** [comment]
Marking this DNM until the CI workflow is fully tested (will get to this some time next week).

File: scripts/release/list_backports.py [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+    def by_date_range(repo, base, start_date, end_date, excludes):
+        """Create a Backport object with the provided repo,
+        base, start datetime object, and end datetime objects, and
+        list of excluded PRs"""
+
+        pulls = []
+
+        unfiltered_pulls = repo.get_pulls(
+            base=base, state='closed')
+        for p in unfiltered_pulls:
```
Comment (fabiobaltieri) [inline scripts/release/list_backports.py:140]: These takes forever for me, are the APIs just that slow? Wondering if it's worth adding a verbose flag to print `unfiltered_pulls.totalCount` here and then a line as PR get fetched, what do you think? (can't see any way of making it faster but maybe I'm missing something)

**Review by fabiobaltieri — DISMISSED** [review]

File: scripts/release/list_backports.py (hunk 11 above)
Comment (fabiobaltieri) [reply scripts/release/list_backports.py:140]: Yeah it does not look like it's possible using that API right now, though `list_issues.py` does that using `requests.get` directly for issues: https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/release/list_issues.py#L109, worth looking into it? Probably not worth it for something we run once per release.

Comment (cfriedt) [reply scripts/release/list_backports.py:140]: It might be worth optimizing that at some point later, or maybe requesting GH add that feature to their API, but yeah, probably not pressing right now.

**Comment by cfriedt** [comment]
> Marking this DNM until the CI workflow is fully tested (will get to this some time next week).

@stephanosio - any update?

File: scripts/release/list_backports.py [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+            # * :github:`38972` - logging: Cleaning references to tracing in logging
+            print(f'* :github:`{i.number}` - {title}')
+
+    def print_json(self):
+        issue_objects = []
+        for i in self.get_issues():
+            obj = {}
+            obj['id'] = i.number
+            obj['title'] = Backport.sanitize_title(i.title)
+            obj['url'] = f'https://github.com/zephyrproject-rtos/zephyr/pull/{i.number}'
```
Comment (stephanosio) [inline scripts/release/list_backports.py:207]: need to use the org and repo names from the arg.

File: scripts/release/list_backports.py [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+        for p in self._pulls:
+            # check for issues in this pr
+            issues_for_this_pr = {}
+            with io.StringIO(p.body) as buf:
+                for line in buf.readlines():
+                    line = line.strip()
+                    match = re.search(r"^Fixes[:]?\s*#([1-9][0-9]*).*", line)
+                    if not match:
+                        match = re.search(
+                            r"^Fixes[:]?\s*https://github\.com/zephyrproject-rtos/zephyr/issues/([1-9][0-9]*).*", line)
```
Comment (stephanosio) [inline scripts/release/list_backports.py:233]: same here

File: scripts/release/list_backports.py [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+                    issue = self._repo.get_issue(issue_number)
+                    if not issue:
+                        if not self._pulls_with_invalid_issues[p.number]:
+                            self._pulls_with_invalid_issues[p.number] = [
+                                issue_number]
+                        else:
+                            self._pulls_with_invalid_issues[p.number].append(
+                                issue_number)
+                        logging.error(
+                            f'https://github.com/zephyrproject-rtos/zephyr/pull/{p.number} references invalid issue number {issue_number}')
```
Comment (stephanosio) [inline scripts/release/list_backports.py:246]: same here

File: scripts/release/list_backports.py [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+                        continue
+                    issues_for_this_pr[issue_number] = issue
+
+            # report prs missing issues later
+            if len(issues_for_this_pr) == 0:
+                logging.error(
+                    f'https://github.com/zephyrproject-rtos/zephyr/pull/{p.number} does not have an associated issue')
```
Comment (stephanosio) [inline scripts/release/list_backports.py:253]: ditto

File: scripts/release/list_backports.py [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+
+    if not bp:
+        return os.EX_DATAERR
+
+    pulls_with_invalid_issues = bp.get_pulls_with_invalid_issues()
+    if pulls_with_invalid_issues:
+        logging.error('The following PRs link to invalid issues:')
+        for (p, lst) in pulls_with_invalid_issues:
+            logging.error(
+                f'\nhttps://github.com/zephyrproject-rtos/zephyr/pull/{p.number}: {lst}')
```
Comment (stephanosio) [inline scripts/release/list_backports.py:318]: ditto

File: scripts/release/list_backports.py [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+        return os.EX_DATAERR
+
+    pulls_without_issues = bp.get_pulls_without_issues()
+    if pulls_without_issues:
+        logging.error(
+            'Please ensure the body of each PR to a release branch contains "Fixes #1234"')
+        logging.error('The following PRs are lacking associated issues:')
+        for p in pulls_without_issues:
+            logging.error(
+                f'https://github.com/zephyrproject-rtos/zephyr/pull/{p.number}')
```
Comment (stephanosio) [inline scripts/release/list_backports.py:328]: ditto

**Review by stephanosio — CHANGES_REQUESTED** [review]
Please update the workflow file as follows:
https://github.com/zephyrproject-rtos/zephyr-testing/blob/v9.9-branch/.github/workflows/backport_issue_check.yml

**Comment by stephanosio** [comment]
@cfriedt also seeing some errors running the script:
```
$ scripts/release/list_backports.py -o zephyrproject-rtos -r zephyr-testing -b v9.9-branch -p 176
Traceback (most recent call last):
  File "scripts/release/list_backports.py", line 340, in <module>
    sys.exit(main())
  File "scripts/release/list_backports.py", line 313, in main
    pulls_with_invalid_issues = bp.get_pulls_with_invalid_issues()
  File "scripts/release/list_backports.py", line 280, in get_pulls_with_invalid_issues
    self.get_issues()
  File "scripts/release/list_backports.py", line 257, in get_issues
    issue_map = issue_map | issues_for_this_pr
TypeError: unsupported operand type(s) for |: 'dict' and 'dict'
```

https://github.com/zephyrproject-rtos/zephyr-testing/pull/176

**Comment by cfriedt** [comment]
> @cfriedt also [seeing some errors running the script](https://github.com/zephyrproject-rtos/zephyr-testing/runs/7362938582?check_suite_focus=true#step:4:11):
> ```
> TypeError: unsupported operand type(s) for |: 'dict' and 'dict'
> ```

@stephanosio - oh wow - is it possible that my python skills are more bleeding edge than Zephyr can handle? (Kidding)

That operator is for appending one dict to another, but I guess only for python 3.9+

https://stackoverflow.com/a/38176082

**Comment by cfriedt** [comment]
> @cfriedt also [seeing some errors running the script](https://github.com/zephyrproject-rtos/zephyr-testing/runs/7362938582?check_suite_focus=true#step:4:11):
> 
> ```
> $ scripts/release/list_backports.py -o zephyrproject-rtos -r zephyr-testing -b v9.9-branch -p 176
> Traceback (most recent call last):
>   File "scripts/release/list_backports.py", line 340, in <module>
>     sys.exit(main())
>   File "scripts/release/list_backports.py", line 313, in main
>     pulls_with_invalid_issues = bp.get_pulls_with_invalid_issues()
>   File "scripts/release/list_backports.py", line 280, in get_pulls_with_invalid_issues
>     self.get_issues()
>   File "scripts/release/list_backports.py", line 257, in get_issues
>     issue_map = issue_map | issues_for_this_pr
> TypeError: unsupported operand type(s) for |: 'dict' and 'dict'
> ```
> 
> [zephyrproject-rtos/zephyr-testing#176](https://github.com/zephyrproject-rtos/zephyr-testing/pull/176)

File: scripts/release/list_backports.py (hunk 17 above)
Comment (cfriedt) [reply scripts/release/list_backports.py:329]: Fixed!

File: scripts/release/list_backports.py (hunk 16 above)
Comment (cfriedt) [reply scripts/release/list_backports.py:319]: Fixed!

File: scripts/release/list_backports.py (hunk 15 above)
Comment (cfriedt) [reply scripts/release/list_backports.py:253]: Fixed!

File: scripts/release/list_backports.py (hunk 14 above)
Comment (cfriedt) [reply scripts/release/list_backports.py:246]: Fixed!

File: scripts/release/list_backports.py (hunk 13 above)
Comment (cfriedt) [reply scripts/release/list_backports.py:233]: Fixed!

File: scripts/release/list_backports.py (hunk 12 above)
Comment (cfriedt) [reply scripts/release/list_backports.py:207]: Fixed!

File: .github/workflows/backport_issue_check.yml [hunk 18]
```diff
@@ -0,0 +1,21 @@
+name: Backport Issue Check
+on:
+  pull_request_target:
+    branches:
+      - v(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)-branch
+      - v(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)-auditable-branch
```
[the remark this answers is not in the document]
Comment (cfriedt) [reply .github/workflows/backport_issue_check.yml:6]: Fixed!

File: .github/workflows/backport_issue_check.yml [hunk 19]
```diff
@@ -0,0 +1,20 @@
+name: Backport Issue Check
```
Comment (stephanosio) [inline .github/workflows/backport_issue_check.yml:1]: Please update the workflow as well: https://github.com/zephyrproject-rtos/zephyr/pull/45880#pullrequestreview-1040626430

Reworked workflow: https://github.com/zephyrproject-rtos/zephyr-testing/blob/v9.9-branch/.github/workflows/backport_issue_check.yml

You can copy and paste the content of the linked workflow file above, which has been verified in the testing repo, into this PR.

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: .github/workflows/backport_issue_check.yml (hunk 19 above)
Comment (cfriedt) [reply .github/workflows/backport_issue_check.yml:1]: Done!

**Review by stephanosio — APPROVED** [review]
Tested in https://github.com/zephyrproject-rtos/zephyr-testing/pull/176

Without linked issue: https://github.com/zephyrproject-rtos/zephyr-testing/runs/7368173127?check_suite_focus=true#step:4:12
With linked issue: https://github.com/zephyrproject-rtos/zephyr-testing/runs/7368167432?check_suite_focus=true#step:4:11

**Comment by cfriedt** [comment]
It's weird that CI still seems to think this PR has a DNM label

**Comment by cfriedt** [comment]
> It's weird that CI still seems to think this PR has a DNM label
> 

Yeah, @stephanosio - it seems removing the DNM label and re-running the job still doesn't pass CI - it seems like kind of a bug, no?

**Comment by stephanosio** [comment]
> Yeah, @stephanosio - it seems removing the DNM label and re-running the job still doesn't pass CI - it seems like kind of a bug, no?

You might need to refresh. For me, "Do Not Merge / Prevent Merging (pull_request)" is shown as skipped.

**Review by fabiobaltieri — APPROVED** [review]

**Comment by stephanosio** [comment]
NOTE: This needs to be "backported" to the relevant release branches to work properly.

v2.7: https://github.com/zephyrproject-rtos/zephyr/pull/47951
v3.0: https://github.com/zephyrproject-rtos/zephyr/pull/47953
v3.1: https://github.com/zephyrproject-rtos/zephyr/pull/47952

**Comment by cfriedt** [comment]
> NOTE: This needs to be "backported" to the relevant release branches to work properly.
> 
> v2.7: #47951 v3.0: #47953 v3.1: #47952

Thanks for merging that into version branches @stephanosio ;-)
